internal/services/network/custom_ip_prefix_resource.go

```diff
@@ -0,0 +1,419 @@
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"signed_message": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"action": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(commissionedActionProvision),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(commissionedActionProvision),
+					string(commissionedActionCommission),
+					string(commissionedActionDecommission),
+					string(commissionedActionDeprovision),
+				}, false),
+			},
+
+			"tags": commonschema.Tags(),
+		},
+	}
+}
+
+func resourceCustomIpPrefixCreateUpdate(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Network.CustomIPPrefixesClient
+	subscriptionId := meta.(*clients.Client).Network.CustomIPPrefixesClient.SubscriptionID
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id := parse.NewCustomIpPrefixID(subscriptionId, d.Get("resource_group_name").(string), d.Get("name").(string))
+
+	state := commissionedAction(d.Get("action").(string))
+
+	existing, err := client.Get(ctx, d.Get("resource_group_name").(string), d.Get("name").(string), "")
+	if d.IsNewResource() {
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+			}
+		}
+
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return tf.ImportAsExistsError("azurerm_custom_ip_prefix", id.ID())
+		}
+	} else {
+		// Since the "Provisioning", "Commissioning", "Decommissioning" or `Deprovisioning` commissioned state of custom ip prefix can do nothing, it could not be updated.
+		if existing.CommissionedState == network.CommissionedStateProvisioning || existing.CommissionedState == network.CommissionedStateCommissioning || existing.CommissionedState == network.CommissionedStateDecommissioning || existing.CommissionedState == network.CommissionedStateDeprovisioning {
+			return fmt.Errorf("the commissioned state of %s is %s, it doesn't allow to be updated", id, string(existing.CommissionedState))
+		}
+
+		switch string(existing.CommissionedState) {
+		case string(network.CommissionedStateProvisioned):
+			if state != commissionedActionCommission && state != commissionedActionDeprovision {
+				return fmt.Errorf("%s can do `Commission` or `Deprovision` when the commissioned state is `Provisioned`", id)
+			}
+		case string(network.CommissionedStateCommissioned):
+			if state != commissionedActionDecommission {
+				return fmt.Errorf("%s can do `Decommission` when the commissioned state is `Commissioned`", id)
+			}
+		case string(commissionedStateDeprovisioned):
+			if state != commissionedActionProvision {
+				return fmt.Errorf("%s can do delete or `Provision` when the commissioned state is `Deprovisioned`", id)
+			}
+		}
```

Comment:
... shouldn't we decommission, update and then recommission here?

Reply:
Thanks for your feedback. For the [commissionedState](https://github.com/Azure/azure-rest-api-specs/blob/1b301846a0e31a668c59b264c900be2128762fac/specification/network/resource-manager/Microsoft.Network/stable/2021-08-01/customIpPrefix.json#L386) property,  there are "Provisioning",   "Provisioned", "Commissioning","Commissioned", "Decommissioning",Deprovisioning" values are listed in the v2021-0-01 API. Per the following information confirmed with service team, I added the check code above. I am not sure what the comment means, so could you please explain about it?

Provisioning: can do nothing  
Commissioning: can do nothing 
Decommissioning: can do nothing
Deprovisioning: can do nothing  
Provisioned: can do commission, or deprovision  
Commissioned: can do decommission  
Deprovisioned: can do delete or provision(backend start from the beginning, do WAN validation again)  

Note: If the action of `Decommission` is done, the commissionState back to Provisioned.

Thanks,
Elena

Reply:
Code has been updated. Could you please see if this is what you mean?